### PR TITLE
Add to spill probe flag to support hash probe spill

### PR DESCRIPTION
**Pull request by xiaoxmeng**
Adds probed flag recording to support probe side spilling. When probe side triggers
hash join table spilling, it needs to records the probed flag of the spilled table for join
type like right join to produce the output. To support this:
(1) spiller for hash join build needs to extract the probed flag from the row container
of the hash join table and spill to disk along with the user table columns. To be specific,
we attach a boolean column at the end of table columns.
(2) hash build side needs the logic to restore the probed flag in the row container of the
rebuilt the hash table based on the value of the attached boolean column read from the
spilled table file.

### CODE REVIEW

File: velox/vector/ComplexVector.h [hunk 1]
```diff
@@ -116,6 +116,9 @@ class RowVector : public BaseVector {
 
   /// Get the child vector at a given offset.
   VectorPtr& childAt(column_index_t index) {
+    if (index >= childrenSize_) {
+      LOG(ERROR) << "bad";
```
Comment (mbasmanova) [inline velox/vector/ComplexVector.h:120]: Left over from debugging?

File: velox/exec/tests/HashJoinBridgeTest.cpp [hunk 2]
```diff
@@ -574,4 +574,19 @@ VELOX_INSTANTIATE_TEST_SUITE_P(
     HashJoinBridgeTest,
     testing::ValuesIn(HashJoinBridgeTest::getTestParams()));
 
+TEST(HashJoinBridgeTest, getHashJoinTableSpillType) {
+  const RowTypePtr tableType = ROW({"k1", "k2"}, {BIGINT(), BIGINT()});
+  const RowTypePtr spillType = hashJoinTableSpillType(tableType);
+  ASSERT_EQ(tableType->size() + 1, spillType->size());
```
Comment (mbasmanova) [inline velox/exec/tests/HashJoinBridgeTest.cpp:580]: Can we create an expected RowType and assert that expected and actual are the same using operator== ?

File: velox/exec/HashJoinBridge.h [hunk 3]
```diff
@@ -169,4 +169,13 @@ bool isHashBuildMemoryPool(const memory::MemoryPool& pool);
 /// Returns true if 'pool' is a hash probe operator's memory pool. The check is
 /// currently based on the pool name.
 bool isHashProbeMemoryPool(const memory::MemoryPool& pool);
+
+/// Returns the type used to spill a given hash table type. The function
+/// attaches a boolean column at the end of 'tableType'. It is used by the hash
+/// join table spilling triggered at the probe side to record if each row has
+/// been probed or not for join type like right join.
+RowTypePtr hashJoinTableSpillType(const RowTypePtr& tableType);
```
Comment (mbasmanova) [inline velox/exec/HashJoinBridge.h:177]: Can we make sure to add an extra column only when needed, e.g. not add it for non-right joins.

File: velox/exec/HashBuild.cpp [hunk 4]
```diff
@@ -203,6 +204,17 @@ void HashBuild::setupSpiller(SpillPartition* spillPartition) {
   if (!spillEnabled()) {
     return;
   }
+  if (spillType_ == nullptr) {
+    spillType_ = hashJoinTableSpillType(tableType_, joinType_);
+    if (needProbedFlagSpill_) {
+      spillProbedFlagChannel_ = spillType_->size() - 1;
+      VELOX_CHECK_NULL(spillProbedFlagVector_);
+      // Creates a constant probed flag vector with all values false for build
```
Comment (mbasmanova) [inline velox/exec/HashBuild.cpp:212]: Why are we setting all values to false? Should we not read these from the RowContainer?

File: velox/core/PlanNode.h [hunk 5]
```diff
@@ -1357,28 +1357,28 @@ FOLLY_ALWAYS_INLINE std::ostream& operator<<(
 enum class JoinType {
   // For each row on the left, find all matching rows on the right and return
   // all combinations.
-  kInner,
+  kInner = 0,
```
Comment (mbasmanova) [inline velox/core/PlanNode.h:1360]: Unrelated change? Why do we need this?

File: velox/exec/tests/SpillerTest.cpp [hunk 6]
```diff
@@ -141,23 +161,28 @@ class SpillerTest : public exec::test::RowContainerTestBase {
...
+         {"float_val", REAL()},
+         {"double_val", DOUBLE()},
+         {"string_val", VARCHAR()},
+         {"array_val", ARRAY(VARCHAR())},
+         {"struct_val",
+          ROW({{"s_int", INTEGER()}, {"s_array", ARRAY(REAL())}})},
+         {"map_val",
+          MAP(VARCHAR(),
+              MAP(BIGINT(),
+                  ROW({{"s2_int", INTEGER()}, {"s2_string", VARCHAR()}})))}});
```
Comment (mbasmanova) [inline velox/exec/tests/SpillerTest.cpp:180]: please, restore previous formatting as it is easier to read

File: velox/exec/tests/HashJoinBridgeTest.cpp [hunk 7]
```diff
@@ -574,4 +574,44 @@ VELOX_INSTANTIATE_TEST_SUITE_P(
     HashJoinBridgeTest,
     testing::ValuesIn(HashJoinBridgeTest::getTestParams()));
 
+TEST(HashJoinBridgeTest, needRightSideJoin) {
```
Comment (mbasmanova) [inline velox/exec/tests/HashJoinBridgeTest.cpp:577]: Not sure how valuable this test it. It seems to repeat the logic in production code. Maybe explicit checks might be better.

```
ASSERT_TRUE(needRightSideJoin(kRight));
ASSERT_TRUE(needRightSideJoin(kFull));
ASSERT_FALSE(needRightSideJoin(kInner));
...
```

File: velox/exec/Spiller.h [hunk 8]
```diff
@@ -39,8 +39,10 @@ class Spiller {
     kOrderByInput = 4,
     // Used for order by output processing stage.
     kOrderByOutput = 5,
+    // Used for row number.
+    kRowNumber = 6,
```
Comment (mbasmanova) [inline velox/exec/Spiller.h:43]: It would be nice to move away from having all spilling logic from all operators be in a single class.

**Review by mbasmanova — APPROVED** [review]

**Review by xiaoxmeng — COMMENTED** [review]
@mbasmanova thanks for the review!
